### PR TITLE
extend/os/linux/dependency_collector.rb: fix circular dependency

### PR DESCRIPTION
**Pull request by danielnachun**
- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

File: Library/Homebrew/extend/os/linux/dependency_collector.rb [hunk 1]
```diff
@@ -63,7 +63,13 @@ def init_global_dep_tree_if_needed!
         end
 
         @@global_dep_tree[global_dep_name].merge(deps.map(&:name))
-        deps = deps.flat_map { |dep| dep.to_formula.deps }
+        deps = deps.flat_map do |dep|
+          # Stop searching if the dependency is glibc because it is the root
+          # of the dependency tree in Linux.
+          next if dep.to_formula.name == GLIBC
+
+          dep.to_formula.deps
```
[the remark this answers is not in the document]
Comment (danielnachun) [reply Library/Homebrew/extend/os/linux/dependency_collector.rb:71]: Indeed that does work - I'll change it.

**Review by carlocab — APPROVED** [review]
Makes sense!

**Comment by Bo98** [comment]
This PR needs a description and steps to reproduce as I do not see any issue locally.

<details>
  <summary>brew deps --tree</summary>

```
$ brew deps --tree --include-build --annotate hello
hello
├── gcc@5 
│   ├── gmp 
│   │   ├── m4  [build]
│   │   │   └── glibc 
│   │   │       └── linux-headers@5.15 
│   │   │           └── rsync  [build]
│   │   │               ├── lz4 
│   │   │               ├── openssl@1.1 
│   │   │               │   └── ca-certificates 
│   │   │               ├── popt 
│   │   │               ├── xxhash 
│   │   │               ├── zstd 
│   │   │               │   ├── cmake  [build]
│   │   │               │   │   ├── ncurses 
│   │   │               │   │   │   ├── pkg-config  [build]
│   │   │               │   │   │   └── gpatch  [build]
│   │   │               │   │   └── openssl@1.1 
│   │   │               │   │       └── ca-certificates 
│   │   │               │   ├── lz4 
│   │   │               │   ├── xz 
│   │   │               │   └── zlib 
│   │   │               │       └── glibc@2.13  [build]
│   │   │               │           └── linux-headers@4.4  [build]
│   │   │               └── zlib 
│   │   │                   └── glibc@2.13  [build]
│   │   │                       └── linux-headers@4.4  [build]
│   │   └── glibc 
│   │       └── linux-headers@5.15 
│   │           └── rsync  [build]
│   │               ├── lz4 
│   │               ├── openssl@1.1 
│   │               │   └── ca-certificates 
│   │               ├── popt 
│   │               ├── xxhash 
│   │               ├── zstd 
│   │               │   ├── cmake  [build]
│   │               │   │   ├── ncurses 
│   │               │   │   │   ├── pkg-config  [build]
│   │               │   │   │   └── gpatch  [build]
│   │               │   │   └── openssl@1.1 
│   │               │   │       └── ca-certificates 
│   │               │   ├── lz4 
│   │               │   ├── xz 
│   │               │   └── zlib 
│   │               │       └── glibc@2.13  [build]
│   │               │           └── linux-headers@4.4  [build]
│   │               └── zlib 
│   │                   └── glibc@2.13  [build]
│   │                       └── linux-headers@4.4  [build]
│   ├── isl@0.18 
│   │   ├── gmp 
│   │   │   ├── m4  [build]
│   │   │   │   └── glibc 
│   │   │   │       └── linux-headers@5.15 
│   │   │   │           └── rsync  [build]
│   │   │   │               ├── lz4 
│   │   │   │               ├── openssl@1.1 
│   │   │   │               │   └── ca-certificates 
│   │   │   │               ├── popt 
│   │   │   │               ├── xxhash 
│   │   │   │               ├── zstd 
│   │   │   │               │   ├── cmake  [build]
│   │   │   │               │   │   ├── ncurses 
│   │   │   │               │   │   │   ├── pkg-config  [build]
│   │   │   │               │   │   │   └── gpatch  [build]
│   │   │   │               │   │   └── openssl@1.1 
│   │   │   │               │   │       └── ca-certificates 
│   │   │   │               │   ├── lz4 
│   │   │   │               │   ├── xz 
│   │   │   │               │   └── zlib 
│   │   │   │               │       └── glibc@2.13  [build]
│   │   │   │               │           └── linux-headers@4.4  [build]
│   │   │   │               └── zlib 
│   │   │   │                   └── glibc@2.13  [build]
│   │   │   │                       └── linux-headers@4.4  [build]
│   │   │   └── glibc 
│   │   │       └── linux-headers@5.15 
│   │   │           └── rsync  [build]
│   │   │               ├── lz4 
│   │   │               ├── openssl@1.1 
│   │   │               │   └── ca-certificates 
│   │   │               ├── popt 
│   │   │               ├── xxhash 
│   │   │               ├── zstd 
│   │   │               │   ├── cmake  [build]
│   │   │               │   │   ├── ncurses 
│   │   │               │   │   │   ├── pkg-config  [build]
│   │   │               │   │   │   └── gpatch  [build]
│   │   │               │   │   └── openssl@1.1 
│   │   │               │   │       └── ca-certificates 
│   │   │               │   ├── lz4 
│   │   │               │   ├── xz 
│   │   │               │   └── zlib 
│   │   │               │       └── glibc@2.13  [build]
│   │   │               │           └── linux-headers@4.4  [build]
│   │   │               └── zlib 
│   │   │                   └── glibc@2.13  [build]
│   │   │                       └── linux-headers@4.4  [build]
│   │   └── glibc 
│   │       └── linux-headers@5.15 
│   │           └── rsync  [build]
│   │               ├── lz4 
│   │               ├── openssl@1.1 
│   │               │   └── ca-certificates 
│   │               ├── popt 
│   │               ├── xxhash 
│   │               ├── zstd 
│   │               │   ├── cmake  [build]
│   │               │   │   ├── ncurses 
│   │               │   │   │   ├── pkg-config  [build]
│   │               │   │   │   └── gpatch  [build]
│   │               │   │   └── openssl@1.1 
│   │               │   │       └── ca-certificates 
│   │               │   ├── lz4 
│   │               │   ├── xz 
│   │               │   └── zlib 
│   │               │       └── glibc@2.13  [build]
│   │               │           └── linux-headers@4.4  [build]
│   │               └── zlib 
│   │                   └── glibc@2.13  [build]
│   │                       └── linux-headers@4.4  [build]
│   ├── libmpc 
│   │   ├── gmp 
│   │   │   ├── m4  [build]
│   │   │   │   └── glibc 
│   │   │   │       └── linux-headers@5.15 
│   │   │   │           └── rsync  [build]
│   │   │   │               ├── lz4 
│   │   │   │               ├── openssl@1.1 
│   │   │   │               │   └── ca-certificates 
│   │   │   │               ├── popt 
│   │   │   │               ├── xxhash 
│   │   │   │               ├── zstd 
│   │   │   │               │   ├── cmake  [build]
│   │   │   │               │   │   ├── ncurses 
│   │   │   │               │   │   │   ├── pkg-config  [build]
│   │   │   │               │   │   │   └── gpatch  [build]
│   │   │   │               │   │   └── openssl@1.1 
│   │   │   │               │   │       └── ca-certificates 
│   │   │   │               │   ├── lz4 
│   │   │   │               │   ├── xz 
│   │   │   │               │   └── zlib 
│   │   │   │               │       └── glibc@2.13  [build]
│   │   │   │               │           └── linux-headers@4.4  [build]
│   │   │   │               └── zlib 
│   │   │   │                   └── glibc@2.13  [build]
│   │   │   │                       └── linux-headers@4.4  [build]
│   │   │   └── glibc 
│   │   │       └── linux-headers@5.15 
│   │   │           └── rsync  [build]
│   │   │               ├── lz4 
│   │   │               ├── openssl@1.1 
│   │   │               │   └── ca-certificates 
│   │   │               ├── popt 
│   │   │               ├── xxhash 
│   │   │               ├── zstd 
│   │   │               │   ├── cmake  [build]
│   │   │               │   │   ├── ncurses 
│   │   │               │   │   │   ├── pkg-config  [build]
│   │   │               │   │   │   └── gpatch  [build]
│   │   │               │   │   └── openssl@1.1 
│   │   │               │   │       └── ca-certificates 
│   │   │               │   ├── lz4 
│   │   │               │   ├── xz 
│   │   │               │   └── zlib 
│   │   │               │       └── glibc@2.13  [build]
│   │   │               │           └── linux-headers@4.4  [build]
│   │   │               └── zlib 
│   │   │                   └── glibc@2.13  [build]
│   │   │                       └── linux-headers@4.4  [build]
│   │   ├── mpfr 
│   │   │   ├── gmp 
│   │   │   │   ├── m4  [build]
│   │   │   │   │   └── glibc 
│   │   │   │   │       └── linux-headers@5.15 
│   │   │   │   │           └── rsync  [build]
│   │   │   │   │               ├── lz4 
│   │   │   │   │               ├── openssl@1.1 
│   │   │   │   │               │   └── ca-certificates 
│   │   │   │   │               ├── popt 
│   │   │   │   │               ├── xxhash 
│   │   │   │   │               ├── zstd 
│   │   │   │   │               │   ├── cmake  [build]
│   │   │   │   │               │   │   ├── ncurses 
│   │   │   │   │               │   │   │   ├── pkg-config  [build]
│   │   │   │   │               │   │   │   └── gpatch  [build]
│   │   │   │   │               │   │   └── openssl@1.1 
│   │   │   │   │               │   │       └── ca-certificates 
│   │   │   │   │               │   ├── lz4 
│   │   │   │   │               │   ├── xz 
│   │   │   │   │               │   └── zlib 
│   │   │   │   │               │       └── glibc@2.13  [build]
│   │   │   │   │               │           └── linux-headers@4.4  [build]
│   │   │   │   │               └── zlib 
│   │   │   │   │                   └── glibc@2.13  [build]
│   │   │   │   │                       └── linux-headers@4.4  [build]
│   │   │   │   └── glibc 
│   │   │   │       └── linux-headers@5.15 
│   │   │   │           └── rsync  [build]
│   │   │   │               ├── lz4 
│   │   │   │               ├── openssl@1.1 
│   │   │   │               │   └── ca-certificates 
│   │   │   │               ├── popt 
│   │   │   │               ├── xxhash 
│   │   │   │               ├── zstd 
│   │   │   │               │   ├── cmake  [build]
│   │   │   │               │   │   ├── ncurses 
│   │   │   │               │   │   │   ├── pkg-config  [build]
│   │   │   │               │   │   │   └── gpatch  [build]
│   │   │   │               │   │   └── openssl@1.1 
│   │   │   │               │   │       └── ca-certificates 
│   │   │   │               │   ├── lz4 
│   │   │   │               │   ├── xz 
│   │   │   │               │   └── zlib 
│   │   │   │               │       └── glibc@2.13  [build]
│   │   │   │               │           └── linux-headers@4.4  [build]
│   │   │   │               └── zlib 
│   │   │   │                   └── glibc@2.13  [build]
│   │   │   │                       └── linux-headers@4.4  [build]
│   │   │   └── glibc 
│   │   │       └── linux-headers@5.15 
│   │   │           └── rsync  [build]
│   │   │               ├── lz4 
│   │   │               ├── openssl@1.1 
│   │   │               │   └── ca-certificates 
│   │   │               ├── popt 
│   │   │               ├── xxhash 
│   │   │               ├── zstd 
│   │   │               │   ├── cmake  [build]
│   │   │               │   │   ├── ncurses 
│   │   │               │   │   │   ├── pkg-config  [build]
│   │   │               │   │   │   └── gpatch  [build]
│   │   │               │   │   └── openssl@1.1 
│   │   │               │   │       └── ca-certificates 
│   │   │               │   ├── lz4 
│   │   │               │   ├── xz 
│   │   │               │   └── zlib 
│   │   │               │       └── glibc@2.13  [build]
│   │   │               │           └── linux-headers@4.4  [build]
│   │   │               └── zlib 
│   │   │                   └── glibc@2.13  [build]
│   │   │                       └── linux-headers@4.4  [build]
│   │   └── glibc 
│   │       └── linux-headers@5.15 
│   │           └── rsync  [build]
│   │               ├── lz4 
│   │               ├── openssl@1.1 
│   │               │   └── ca-certificates 
│   │               ├── popt 
│   │               ├── xxhash 
│   │               ├── zstd 
│   │               │   ├── cmake  [build]
│   │               │   │   ├── ncurses 
│   │               │   │   │   ├── pkg-config  [build]
│   │               │   │   │   └── gpatch  [build]
│   │               │   │   └── openssl@1.1 
│   │               │   │       └── ca-certificates 
│   │               │   ├── lz4 
│   │               │   ├── xz 
│   │               │   └── zlib 
│   │               │       └── glibc@2.13  [build]
│   │               │           └── linux-headers@4.4  [build]
│   │               └── zlib 
│   │                   └── glibc@2.13  [build]
│   │                       └── linux-headers@4.4  [build]
│   ├── mpfr 
│   │   ├── gmp 
│   │   │   ├── m4  [build]
│   │   │   │   └── glibc 
│   │   │   │       └── linux-headers@5.15 
│   │   │   │           └── rsync  [build]
│   │   │   │               ├── lz4 
│   │   │   │               ├── openssl@1.1 
│   │   │   │               │   └── ca-certificates 
│   │   │   │               ├── popt 
│   │   │   │               ├── xxhash 
│   │   │   │               ├── zstd 
│   │   │   │               │   ├── cmake  [build]
│   │   │   │               │   │   ├── ncurses 
│   │   │   │               │   │   │   ├── pkg-config  [build]
│   │   │   │               │   │   │   └── gpatch  [build]
│   │   │   │               │   │   └── openssl@1.1 
│   │   │   │               │   │       └── ca-certificates 
│   │   │   │               │   ├── lz4 
│   │   │   │               │   ├── xz 
│   │   │   │               │   └── zlib 
│   │   │   │               │       └── glibc@2.13  [build]
│   │   │   │               │           └── linux-headers@4.4  [build]
│   │   │   │               └── zlib 
│   │   │   │                   └── glibc@2.13  [build]
│   │   │   │                       └── linux-headers@4.4  [build]
│   │   │   └── glibc 
│   │   │       └── linux-headers@5.15 
│   │   │           └── rsync  [build]
│   │   │               ├── lz4 
│   │   │               ├── openssl@1.1 
│   │   │               │   └── ca-certificates 
│   │   │               ├── popt 
│   │   │               ├── xxhash 
│   │   │               ├── zstd 
│   │   │               │   ├── cmake  [build]
│   │   │               │   │   ├── ncurses 
│   │   │               │   │   │   ├── pkg-config  [build]
│   │   │               │   │   │   └── gpatch  [build]
│   │   │               │   │   └── openssl@1.1 
│   │   │               │   │       └── ca-certificates 
│   │   │               │   ├── lz4 
│   │   │               │   ├── xz 
│   │   │               │   └── zlib 
│   │   │               │       └── glibc@2.13  [build]
│   │   │               │           └── linux-headers@4.4  [build]
│   │   │               └── zlib 
│   │   │                   └── glibc@2.13  [build]
│   │   │                       └── linux-headers@4.4  [build]
│   │   └── glibc 
│   │       └── linux-headers@5.15 
│   │           └── rsync  [build]
│   │               ├── lz4 
│   │               ├── openssl@1.1 
│   │               │   └── ca-certificates 
│   │               ├── popt 
│   │               ├── xxhash 
│   │               ├── zstd 
│   │               │   ├── cmake  [build]
│   │               │   │   ├── ncurses 
│   │               │   │   │   ├── pkg-config  [build]
│   │               │   │   │   └── gpatch  [build]
│   │               │   │   └── openssl@1.1 
│   │               │   │       └── ca-certificates 
│   │               │   ├── lz4 
│   │               │   ├── xz 
│   │               │   └── zlib 
│   │               │       └── glibc@2.13  [build]
│   │               │           └── linux-headers@4.4  [build]
│   │               └── zlib 
│   │                   └── glibc@2.13  [build]
│   │                       └── linux-headers@4.4  [build]
│   ├── zlib 
│   │   └── glibc@2.13  [build]
│   │       └── linux-headers@4.4  [build]
│   ├── binutils 
│   │   ├── zlib 
│   │   │   └── glibc@2.13  [build]
│   │   │       └── linux-headers@4.4  [build]
│   │   ├── glibc@2.13  [build]
│   │   │   └── linux-headers@4.4  [build]
│   │   ├── linux-headers@4.4  [build]
│   │   └── glibc 
│   │       └── linux-headers@5.15 
│   │           └── rsync  [build]
│   │               ├── lz4 
│   │               ├── openssl@1.1 
│   │               │   └── ca-certificates 
│   │               ├── popt 
│   │               ├── xxhash 
│   │               ├── zstd 
│   │               │   ├── cmake  [build]
│   │               │   │   ├── ncurses 
│   │               │   │   │   ├── pkg-config  [build]
│   │               │   │   │   └── gpatch  [build]
│   │               │   │   └── openssl@1.1 
│   │               │   │       └── ca-certificates 
│   │               │   ├── lz4 
│   │               │   ├── xz 
│   │               │   └── zlib 
│   │               │       └── glibc@2.13  [build]
│   │               │           └── linux-headers@4.4  [build]
│   │               └── zlib 
│   │                   └── glibc@2.13  [build]
│   │                       └── linux-headers@4.4  [build]
│   ├── glibc 
│   │   └── linux-headers@5.15 
│   │       └── rsync  [build]
│   │           ├── lz4 
│   │           ├── openssl@1.1 
│   │           │   └── ca-certificates 
│   │           ├── popt 
│   │           ├── xxhash 
│   │           ├── zstd 
│   │           │   ├── cmake  [build]
│   │           │   │   ├── ncurses 
│   │           │   │   │   ├── pkg-config  [build]
│   │           │   │   │   └── gpatch  [build]
│   │           │   │   └── openssl@1.1 
│   │           │   │       └── ca-certificates 
│   │           │   ├── lz4 
│   │           │   ├── xz 
│   │           │   └── zlib 
│   │           │       └── glibc@2.13  [build]
│   │           │           └── linux-headers@4.4  [build]
│   │           └── zlib 
│   │               └── glibc@2.13  [build]
│   │                   └── linux-headers@4.4  [build]
│   └── glibc 
│       └── linux-headers@5.15 
│           └── rsync  [build]
│               ├── lz4 
│               ├── openssl@1.1 
│               │   └── ca-certificates 
│               ├── popt 
│               ├── xxhash 
│               ├── zstd 
│               │   ├── cmake  [build]
│               │   │   ├── ncurses 
│               │   │   │   ├── pkg-config  [build]
│               │   │   │   └── gpatch  [build]
│               │   │   └── openssl@1.1 
│               │   │       └── ca-certificates 
│               │   ├── lz4 
│               │   ├── xz 
│               │   └── zlib 
│               │       └── glibc@2.13  [build]
│               │           └── linux-headers@4.4  [build]
│               └── zlib 
│                   └── glibc@2.13  [build]
│                       └── linux-headers@4.4  [build]
└── glibc 
    └── linux-headers@5.15 
        └── rsync  [build]
            ├── lz4 
            ├── openssl@1.1 
            │   └── ca-certificates 
            ├── popt 
            ├── xxhash 
            ├── zstd 
            │   ├── cmake  [build]
            │   │   ├── ncurses 
            │   │   │   ├── pkg-config  [build]
            │   │   │   └── gpatch  [build]
            │   │   └── openssl@1.1 
            │   │       └── ca-certificates 
            │   ├── lz4 
            │   ├── xz 
            │   └── zlib 
            │       └── glibc@2.13  [build]
            │           └── linux-headers@4.4  [build]
            └── zlib 
                └── glibc@2.13  [build]
                    └── linux-headers@4.4  [build]
```
</summary>

**Review by MikeMcQuaid — APPROVED** [review]
I'm fine with this approach temporarily if it unblocks us. I really want to cleanup this code for 3.6.0, though, as it's rapidly becoming incomprehensible.

**Review by Bo98 — CHANGES_REQUESTED** [review]
This introduces circular dependencies rather than fix any.

```
hello
├── gcc@5 
│   ├── gmp 
│   │   ├── m4  [build]
│   │   │   ├── gcc@5  (CIRCULAR DEPENDENCY)
│   │   │   └── glibc 
│   │   │       └── linux-headers@5.15 
│   │   │           └── rsync  [build]
│   │   │               ├── lz4 
│   │   │               ├── openssl@1.1 
│   │   │               │   └── ca-certificates 
│   │   │               ├── popt 
│   │   │               ├── xxhash 
│   │   │               ├── zstd 
│   │   │               │   ├── cmake  [build]
│   │   │               │   │   ├── ncurses 
│   │   │               │   │   │   ├── pkg-config  [build]
│   │   │               │   │   │   └── gpatch  [build]
│   │   │               │   │   └── openssl@1.1 
│   │   │               │   │       └── ca-certificates 
│   │   │               │   ├── lz4 
│   │   │               │   ├── xz 
│   │   │               │   └── zlib 
│   │   │               │       └── glibc@2.13  [build]
│   │   │               │           └── linux-headers@4.4  [build]
│   │   │               └── zlib 
│   │   │                   └── glibc@2.13  [build]
│   │   │                       └── linux-headers@4.4  [build]
...
```

**Comment by danielnachun** [comment]
Hmm, I do see the circular dependencies now.  One detail I forgot to mention above - I think this problem only occurs when brewed `glibc` in installed, so you have to test it in an old enough system like CentOS 7 where brewed `glibc` is currently needed to see the failure.

Although the circular dependency shown above does occur, it doesn't actually cause the build to fail. Though we still need to figure out its origin.  The change I made shouldn't be adding any dependencies, only removing them.

**Comment by danielnachun** [comment]
https://github.com/Homebrew/homebrew-core/pull/108974/files is a better solution to this problem.  It was not clear initially if the excessive tree depth was due to an infinite loop, but it wasn't.  So we can close this as it's not fixing a real problem.

**Pull request closed without merging**